src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs

```diff
@@ -0,0 +1,76 @@
+        /// Assets information can be null and returns the package from the project files.
+        /// Checks if package already exists in the project and return it, otherwise update the project installed packages.
+        /// </summary>
+        /// <param name="projectLibrary">Library from the project file.</param>
+        /// <param name="targetFramework">Target framework from the project file.</param>
+        /// <param name="assetsTargetFrameworkInformation">Target framework information from the assets file.</param>
+        /// <param name="targets"></param>
+        /// <param name="installedPackages">Installed packages information from the project.</param>
+        internal static PackageIdentity UpdateResolvedVersion(LibraryDependency projectLibrary, NuGetFramework targetFramework, TargetFrameworkInformation assetsTargetFrameworkInformation, IList<LockFileTarget> targets, Dictionary<string, ProjectInstalledPackage> installedPackages)
+        {
+            NuGetVersion resolvedVersion = default;
+
+            // Returns the installed version if the package:
+            // 1. Already exists in the installedPackages
+            // 2. The range is the same as the installed one
+            // 3. There are no changes in the assets file
+            if (installedPackages.TryGetValue(projectLibrary.Name, out ProjectInstalledPackage installedVersion) && installedVersion.AllowedVersions.Equals(projectLibrary.LibraryRange.VersionRange) && targets == null)
+            {
+                return installedVersion.InstalledPackage;
+            }
+
+            resolvedVersion = GetInstalledVersion(projectLibrary, targetFramework, assetsTargetFrameworkInformation, targets);
+
+            if (resolvedVersion == null)
+            {
+                resolvedVersion = projectLibrary.LibraryRange?.VersionRange?.MinVersion ?? new NuGetVersion(0, 0, 0);
+            }
+
+            // Add or update the the version of the package in the project
+            if (installedPackages.TryGetValue(projectLibrary.Name, out ProjectInstalledPackage installedPackage))
+            {
+                installedPackages[projectLibrary.Name] = new ProjectInstalledPackage(projectLibrary.LibraryRange.VersionRange, new PackageIdentity(projectLibrary.Name, resolvedVersion));
+            }
+            else
+            {
+                ProjectInstalledPackage newInstalledPackage = new ProjectInstalledPackage(projectLibrary.LibraryRange.VersionRange, new PackageIdentity(projectLibrary.Name, resolvedVersion));
+                installedPackages.Add(projectLibrary.Name, newInstalledPackage);
+            }
+
+            return new PackageIdentity(projectLibrary.Name, resolvedVersion);
+        }
+
+        private static NuGetVersion GetInstalledVersion(LibraryDependency libraryProjectFile, NuGetFramework targetFramework, TargetFrameworkInformation assetsTargetFrameworkInformation, IList<LockFileTarget> targets)
```

Comment:
This is a FYI to think about rather than request for change: I once read advice to be specific about the type you return and be liberal about the types you accept. I really liked that advice.

The `targets` parameter is currently defined as `IList<LockFileTargets>` where `IList<T>` extends `ICollection<T>` with an indexer operator (`[int] {get; set; }`), and `ICollection<T>` extends `IEnumerable<T>` with a `.Count` property. However, this method doesn't use either `.Count` or the indexer, therefore this method could just accept an `IEnumerable<T>`, so that it accept a wider range of inputs without changing the method body at all.

In other cases where you do want to get the `.Count` or use an indexer, absolutely do use the more specific interfaces. Don't use a less efficient algorithm just to force `IEnumerable<T>` everywhere. But when you don't need something more specific, being less specific might allow the method to be re-used more easily, without having to allocate new lists/arrays.